### ISSUE TITLE
Buf compiled with protobuf v1.4.0 drops Custom Options

### ISSUE THREAD

**Issue by robsonmeemo** [bufbuild/buf#52]
```
mkdir foo
cd foo
go mod init foo.bar
go get google.golang.org/protobuf/proto
go get github.com/bufbuild/buf/cmd/buf
```

The binary installed will drop any custom options defined in your proto files when generating the Image file.

**Comment by bufdev** [comment]
This is not too surprising - we are not upgraded to google.golang.org/protobuf yet in Buf, but will be soon. We will make sure to check this works before pushing Buf with upgrades.

**Comment by jhump** [comment]
@robsonmeemo, can you provide a little more detail? I assume if it drops all custom options, then it should be trivial to create a repro case. How are you inspecting the resulting image? Can you provide a little more detail as to how you encountered this?

The core proto-parsing logic in buf comes from github.com/jhump/protoreflect/desc/protoparse. And all of the tests for that stuff pass with the latest proto runtime, so I'm curious for more details on exactly what you are seeing.

**Comment by robsonmeemo** [comment]
```
mkdir foo
cd foo
go mod init foo.bar
go get google.golang.org/protobuf/proto
go get github.com/bufbuild/buf/cmd/buf
```

```
% cat a.proto
syntax = "proto3";

import "google/protobuf/descriptor.proto";

extend google.protobuf.FieldOptions {
  int32 baz = 50007;
}

message Foo {
  string bar = 1 [(baz) = 42];
}
```

```
% cat image.proto
syntax = "proto3";

import "google/protobuf/descriptor.proto";

message Image {
  // file matches the file field of a FileDescriptorSet.
  repeated google.protobuf.FileDescriptorProto file = 1;

  // bufbuild_image_extension is the ImageExtension for this image.
  //
  // The prefixed name and high tag value is used to all but guarantee there
  // will never be any conflict with Google's FileDescriptorSet definition.
  // The definition of a FileDescriptorSet has not changed in 11 years, so
  // we're not too worried about a conflict here.
 // optional ImageExtension bufbuild_image_extension = 8042;
}
```

```
% buf image build -o a.bin
% cat a.bin | protoc --decode Image image.proto | grep -A 9 'name: "Foo"'
    name: "Foo"
    field {
      name: "bar"
      number: 1
      label: LABEL_OPTIONAL
      type: TYPE_STRING
      options {
      }
      json_name: "bar"
    }
```

**Comment by bufdev** [comment]
I added `--exclude-imports --exclude-source-info` to `buf image build` to make this cleaner.

Reproduction above (using new golang/protobuf):

```
$ cat a-reproduction.bin | protoc --decode Image image.proto
file {
  name: "a.proto"
  dependency: "google/protobuf/descriptor.proto"
  message_type {
    name: "Foo"
    field {
      name: "bar"
      number: 1
      label: LABEL_OPTIONAL
      type: TYPE_STRING
      options {
      }
      json_name: "bar"
    }
  }
  extension {
    name: "baz"
    extendee: ".google.protobuf.FieldOptions"
    number: 50007
    label: LABEL_OPTIONAL
    type: TYPE_INT32
    json_name: "baz"
  }
  syntax: "proto3"
}
```

Same thing with released version of `buf` (which uses old golang/protobuf right now):

```
$ cat a-release.bin | protoc --decode Image image.proto
file {
  name: "a.proto"
  dependency: "google/protobuf/descriptor.proto"
  message_type {
    name: "Foo"
    field {
      name: "bar"
      number: 1
      label: LABEL_OPTIONAL
      type: TYPE_STRING
      options {
        50007: 42
      }
      json_name: "bar"
    }
  }
  extension {
    name: "baz"
    extendee: ".google.protobuf.FieldOptions"
    number: 50007
    label: LABEL_OPTIONAL
    type: TYPE_INT32
    json_name: "baz"
  }
  syntax: "proto3"
}
```

Building a `FileDescriptorSet` using `protoc -I . -o a-protoc.bin a.proto`:

```
$ cat a-protoc.bin | protoc --decode Image image.proto
file {
  name: "a.proto"
  dependency: "google/protobuf/descriptor.proto"
  message_type {
    name: "Foo"
    field {
      name: "bar"
      number: 1
      label: LABEL_OPTIONAL
      type: TYPE_STRING
      options {
        50007: 42
      }
      json_name: "bar"
    }
  }
  extension {
    name: "baz"
    extendee: ".google.protobuf.FieldOptions"
    number: 50007
    label: LABEL_OPTIONAL
    type: TYPE_INT32
    json_name: "baz"
  }
  syntax: "proto3"
}
```

Diff from reproduction to release (release/protoc are the same):

```
$ diff a-reproduction.txt a-release.txt
11a12
>         50007: 42
```

So it does appear to drop.

This may be an internal thing with buf as well. Will investigate. I have a branch open to upgrade to the new golang/protobuf internally, there's just some tricky stuff involved around `FileDescriptorSets` I want to make sure we get right.

**Comment by robsonmeemo** [comment]
Does this look hard to fix? I'm considering whether to work around the bug on our side or wait for you guys to fix it.

**Comment by bufdev** [comment]
We have a bunch of things on our plate right now, we will get to this though.

What is need for this to be fixed right now? The version is locked with the go.mod, so this shouldn't break you, and all the code is in `internal/` so there's no exposed library.

**Comment by robsonmeemo** [comment]
We currently manage our go binaries using go mod. We want to migrate our main code to the new protobuf library but buf breaks when we do it. We'll have to special case buf to be the only go binary not managed through go mod. It's doable but I'd like to avoid that if this bug was planned to be fixed in the next couple of days.

**Comment by bufdev** [comment]
I'm guessing you mean that you have something in your code (monorepo?) such as:

```go
import _ "github.com/bufbuid/buf/cmd/buf"
```

Is this accurate?

We'd really recommend installing binaries via the released binaries, or if you want to use the source, to do so via `GOBIN=/path/to/bin go get github.com/bufbuild/buf/cmd/buf@VERSION`. If you don't want `go get` to touch your mainline go.mod, then `cd` to a temp directory first such as in https://github.com/bufbuild/makego/blob/master/make/go/dep_staticcheck.mk#L17 - this is generally a good idea for ALL go binaries you install from source for this exact reason, to make sure these binaries are built with the exact library versions they expect.

Regardless, we'll try to get this fixed ASAP because we need to anyways, but I can't promise next couple of days.

**Comment by robsonmeemo** [comment]
My go.mod pins buf to v0.11.0.
```
github.com/bufbuild/buf v0.11.0
```

We do take care to download the exact version that is supposed to be supported:
```
CURRENT_BUF_VERSION=$(go list -m github.com/bufbuild/buf | cut -d" " -f2)
go get github.com/bufbuild/buf/cmd/buf@$CURRENT_BUF_VERSION
```

The issue is that upgrading protobuf makes buf v0.11.0 use the newer dependency and break.

**Comment by bufdev** [comment]
If you do the above, it will not upgrade the version, specifically:

```
TMP=$(mktemp -d)
cd ${TMP} && go get github.com/bufbuild/buf/cmd/buf@v0.11.0 && cd - && rm -rf ${TMP}
```

The only way this would download the version of golang/protobuf is if you have an existing go.mod controlling the `go get` call. Otherwise, golang/protobuf will not be upgraded.

This is how we'd recommend doing all golang binary installs - if you are calling `go get` in a directory that has a `go.mod`, then this could be affected (as you are likely affected with this).

**Comment by robsonmeemo** [comment]
I'm following the recommended way by the Go team:
https://github.com/go-modules-by-example/index/blob/master/010_tools/README.md

**Comment by bufdev** [comment]
We wouldn't recommend that, for the above reason - you'll end up with breaking issues in your golang binaries (not just buf). This is why we do the above in our setup, to make sure this does not happen. In our experience, breaking changes do happen on minors (for example, 1.3.5 -> 1.4.0 has some edge cases of incompatibility, as this issue describes), so we always want to make sure binaries are built with the exact versions of the libraries they expect.

**Comment by robsonmeemo** [comment]
Thank you. It does make sense. Too bad the go team is passing around bad advice. :( I'm no longer blocked by this bug.

**Comment by bufdev** [comment]
@robsonmeemo see https://github.com/bufbuild/buf/pull/53 - haven't tested this issue yet, but will now

**Comment by bufdev** [comment]
Can confirm this is an issue when fully migrated (ie the option is dropped)

**Comment by bufdev** [comment]
@jhump I've reproduced the issue in a test.

https://github.com/bufbuild/buf/blob/87edc2dba44a049db4dd0b4630607f842879193b/internal/buf/bufbuild/testdata/customoptions1/a.proto

With golang/protobuf v1 (master branch on buf):

https://github.com/bufbuild/buf/blob/87edc2dba44a049db4dd0b4630607f842879193b/internal/buf/bufbuild/builder_test.go#L109

With golang/protobuf v2 (proto-v2 branch on buf):

https://github.com/bufbuild/buf/blob/8cc109b03b6e6f03a6a0ac204610bfdcc8af3b11/internal/buf/bufbuild/builder_test.go#L111

It works on v1, not on v2 - need to do more investigation. Note the test is skipped on proto-v2 for now, but this does at least show that this is dropped somehow with v2.

It may be something to do with `XXX_unrecognized`, as this is no longer a field for golang/protobuf v2 generated structs (they now have private variables for extensions/unknown fields).

**Comment by bufdev** [comment]
Note there is no direct equivalent to `proto.RegisterExtension` in golang/protobuf v2, working on it

**Comment by bufdev** [comment]
So it's clear what the diff is:

```diff
diff --git a/internal/buf/bufbuild/builder_test.go b/internal/buf/bufbuild/builder_test.go
index 5bee0f3..44a2d18 100644
--- a/internal/buf/bufbuild/builder_test.go
+++ b/internal/buf/bufbuild/builder_test.go
@@ -32,11 +32,12 @@ import (
        "github.com/bufbuild/buf/internal/pkg/storage/storagepath"
        "github.com/bufbuild/buf/internal/pkg/util/utilgithub/utilgithubtesting"
        "github.com/bufbuild/buf/internal/pkg/util/utilproto/utilprototesting"
-       "github.com/golang/protobuf/proto"
-       "github.com/golang/protobuf/protoc-gen-go/descriptor"
        "github.com/stretchr/testify/assert"
        "github.com/stretchr/testify/require"
        "go.uber.org/zap"
+       "google.golang.org/protobuf/proto"
+       "google.golang.org/protobuf/runtime/protoimpl"
+       "google.golang.org/protobuf/types/descriptorpb"
 )

 const (
@@ -44,8 +45,9 @@ const (
 )

 var (
-       testEBaz = &proto.ExtensionDesc{
-               ExtendedType:  (*descriptor.FieldOptions)(nil),
+       testEBaz = &protoimpl.ExtensionInfo{
+               ExtendedType:  (*descriptorpb.FieldOptions)(nil),
                ExtensionType: (*int32)(nil),
                Field:         50007,
                Name:          "baz",
@@ -56,9 +58,9 @@ var (
        testLock              sync.Mutex
 )

-func init() {
-       proto.RegisterExtension(testEBaz)
-}

 func TestGoogleapis(t *testing.T) {
        t.Parallel()
@@ -108,6 +110,7 @@ func TestCompareGoogleapis(t *testing.T) {

 func TestCustomOptions(t *testing.T) {
        // https://github.com/bufbuild/buf/issues/52
+       t.Skip()

        t.Parallel()
        image, fileAnnotations := testBuild(t, false, false, filepath.Join("testdata", "customoptions1"))
@@ -123,8 +126,7 @@ func TestCustomOptions(t *testing.T) {
        fieldOptions := fieldDescriptorProto.Options
        require.NotNil(t, fieldOptions)
        require.True(t, proto.HasExtension(fieldOptions, testEBaz))
-       value, err := proto.GetExtension(fieldOptions, testEBaz)
-       require.NoError(t, err)
+       value := proto.GetExtension(fieldOptions, testEBaz)
        valueInt32, ok := value.(*int32)
        require.True(t, ok)
        require.NotNil(t, valueInt32)
@@ -214,7 +216,7 @@ func testBuildGoogleapis(t *testing.T, includeSourceInfo bool) *imagev1beta1.Ima
        return image
 }
```

The master branch works, the proto-v2 branch fails at `require.True(t, proto.HasExtension(fieldOptions, testEBaz))`

**Comment by bufdev** [comment]
@jhump I could be completely wrong, but I think it comes down to https://github.com/jhump/protoreflect/blob/master/desc/protoparse/options.go#L924 and how the `dynamic` package handles this...this might be a huge issue with v1/v2

**Comment by bufdev** [comment]
Update: it's not a protoreflect issue, see https://github.com/golang/protobuf/issues/1120

**Comment by jhump** [comment]
@bufdev, I think there actually _could_ be an issue in protoreflect that is causing this. Take a look at https://github.com/jhump/protoreflect/pull/319 and see if it helps.

I don't believe binary marshaling needs to know anything about extensions because it stores unrecognized fields in an already marshaled form. So it just concatenates any unrecognized data to the serialized form. But, if the options messages in the new Go runtime (e.g. the compiled Go code for `google.protobuf.MessageOptions`) do _not_ have `XXX_unrecognized` fields, that could totally caused them to be dropped by protoparse.

**Comment by bufdev** [comment]
Yep, you fixed it, haha. I checked that this all is fixed with that branch :-)

**Comment by bufdev** [comment]
This should be fixed now on 0.12.0.
